specs/core/0_beacon-chain.md

```diff
@@ -999,130 +990,161 @@ The output of `get_block_hash` should not change, except that it will no longer
+For each `special` in `block.specials`:
 
-For each `SpecialRecord` `obj` in `block.specials`, verify that its `kind` is one of the below values, and that `obj.data` deserializes according to the format for the given `kind`, then process it. The word "verify" when used below means that if the given verification check fails, the block containing that `SpecialRecord` is invalid.
+* Verify that `special.kind` is a valid values.
+* Verify that `special.data` deserializes according to the format for the given `kind`.
+* Process `special.data` as specified below for each `kind`.
 
-#### LOGOUT
+#### `VOLUNTARY_EXIT`
 
 ```python
 {
+    'slot': 'unit64',
     'validator_index': 'uint64',
-    'signature': '[uint384]'
+    'signature': '[uint384]',
 }
 ```
-Perform the following checks:
-* Verify that `BLSVerify(pubkey=validators[data.validator_index].pubkey, msg=bytes([0] * 32), sig=data.signature, domain=get_domain(state.fork_data, current_slot, DOMAIN_LOGOUT))`.
-* Verify that `validators[validator_index].status == ACTIVE`.
-* Verify that `block.slot >= last_status_change_slot + SHARD_PERSISTENT_COMMITTEE_CHANGE_PERIOD`.
 
-Run `exit_validator(data.validator_index, state, block, penalize=False, current_slot=block.slot)`.
+* Let `validator = state.validator_registry[special.data.validator_index]`.
+* Verify that `BLSVerify(pubkey=validator.pubkey, msg=ZERO_HASH, sig=special.data.signature, domain=get_domain(state.fork_data, special.data.slot, DOMAIN_EXIT))`.
+* Verify that `validator.status == ACTIVE`.
+* Verify that `block.slot >= special.data.slot`.
+* Verify that `block.slot >= validator.last_status_change_slot + SHARD_PERSISTENT_COMMITTEE_CHANGE_PERIOD`.
+* Run `exit_validator(special.data.validator_index, state, penalize=False, current_slot=block.slot)`.
 
-#### CASPER_SLASHING
+#### `CASPER_SLASHING`
 
-```python
+We define the following `SpecialAttestationData` object and the helper `verify_special_attestation_data`:
+ 
+ ```python
 {
-    'vote1_aggregate_sig_indices': '[uint24]',
-    'vote1_data': AttestationSignedData,
-    'vote1_aggregate_sig': '[uint384]',
-    'vote2_aggregate_sig_indices': '[uint24]',
-    'vote2_data': AttestationSignedData,
-    'vote2_aggregate_sig': '[uint384]',
+    'aggregate_sig_poc_0_indices': '[uint24]',
+    'aggregate_sig_poc_1_indices': '[uint24]',
+    'data': AttestationData,
+    'aggregate_sig': '[uint384]',
 }
 ```
 
-Perform the following checks:
+ ```python
+def verify_special_attestation_data(state: State, obj: SpecialAttestationData) -> bool:
+    pubs = [aggregate_pubkey([state.validators[i].pubkey for i in obj.aggregate_sig_poc_0_indices]),
+            aggregate_pubkey([state.validators[i].pubkey for i in obj.aggregate_sig_poc_1_indices])]
+    return BLSMultiVerify(pubkeys=pubs, msgs=[SSZTreeHash(obj)+bytes1(0), SSZTreeHash(obj)+bytes1(1), sig=aggregate_sig)
+```
```

Comment:
The reason why I put `verify_special_attestation_data` outside of the CASPER_SLASHING subsection is that the same method will in phase 1 be used for proof of custody challenges.